### DISCUSSION TITLE
Save configuration in project.yaml instead of database

### DISCUSSION THREAD

**Discussion by timmyomahony** [timmyomahony/craft-matrix-field-preview#114] [Ideas]
Currently all configuration is stored in the database via custom tables and not the `project.yaml`. Although this might go against common convention with Craft, there's a reason for it.

In my own workflow, my local and staging environments are usually separate to my production environment. There may be different content due to testing features, different fields and assets etc. This means that I don't want to store my matrix field preview configuration in the `project.yaml` as - for example - the assets that are configured there are not present in my production asset folder.

That said, there have been a number of requests for this change, where the plugin moves from storing configuration in the `project.yaml` instead of the database.

- https://github.com/weareferal/craft-matrix-field-preview/discussions/76
- https://github.com/weareferal/craft-matrix-field-preview/issues/102

I'd like to hear more on how people use the plugin between environments, as making this change would not be backward compatible.

**Comment by turnstylerj** [comment]
I definitely vote for this change. Craft recommends disabling admin changes in production environments, so there's currently no way to adjust settings for newly-added fields without temporarily enabling admin changes in prod.

**Comment by johnwbaxter** [comment]
We really like the approach MatrixMate has where it is configured in it's own config file which we can add logic into if need be.

It would be great if Matrix Preview followed the same approach!

**Comment by andyharris** [comment]
+1 for this change, would fit our workflow perfectly!

**Comment by martijnvanonz** [comment]
Hi all,

I’m considering the plugin for our clients, but saving data in the database is a bit of a dealbreaker. Any updates on this?
I also wouldn’t mind if the YAML contained absolute paths to the images.

**Comment by timmyomahony** [comment]
Just an update to mention that I'm not ignoring this feature request.

Changing from custom database tables to using the project config would require a fundamental refactoring of the plugin, so I want to be careful that it makes sense to do it. I would like to continue getting some feedback on the reasons for wanting this feature.

## Editing previews in production

My understanding to date has been that the main complaint with having the configuration in the database has been that *you can't easily configure the plugin in production*.

👉 This isn't actually true, **you should always be able to configure your previews in the settings page, even in production**.

This has been poorly communicated by me in the documentation. Furthermore, there was also a bug that meant the plugin icon wasn't showing up in the settings page in Craft 5 when `allowAdminChanges` was `false`, adding to the confusion.

So to make this clearer:

- I've released an update [`5.2.0`](https://github.com/timmyomahony/craft-matrix-field-preview/releases/tag/5.2.0) to the plugin that fixed the issue with the plugin icon so that it's always there (Craft `5.6` or greater is required for this update).
- I've updated the documentation to include [a "deployment" section](https://craft-plugins.timmyomahony.com/matrix-field-preview/docs/deployment) that makes this clearer. 

## Should we still switch to project config?

With the above clarification in mind, should the plugin still switch to using the project config? What other reasons am I overlooking?

**Reply by turnstylerj** [reply]
Depending on the version of Craft 5, the settings page isn't accessible at all if `allowAdminChanges` is set to `false`. The read-only view was added in [5.6.0](https://github.com/craftcms/cms/issues/4371#issuecomment-2605535480).

**Reply by timmyomahony** [reply]
Thanks for the heads-up. I've pushed a new release with a constraint on the Craft version to require 5.6. I've also added some context in release notes to indicate the mistake.

**Comment by idleog** [comment]
There must be a solution in order to be able to deploy config changes in dev environment to staging or production.

**Comment by dcplebranch** [comment]
Just adding my vote for this feature. It would be great if I could just deploy the site and the matrix preview images worked automatically instead of having to go in the production site as admin and configure everything on there separately.
